### PR TITLE
#170 new list control

### PR DESCRIPTION
**Pull request by caritaou**
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

**Comment by matthoward366** [comment]
@caritaou This seems wrong.
Screen Shot 2020-07-21 at 8 38 24 AM

On hover over the row looks different from other table.  The background for the integer/text control is white but when those controls are in a table they show up gray.

**Comment by caritaou** [comment]
@matthoward366 I removed the style overrides to have the same look and feel as other controls.

For the table example, I did try this out but it doesn't work at the moment. I think we need to add support for complex structures within complex structures before this can work. We're not able to update a list item within a subcontrol because the propertyId does not go down that far

**Comment by matthoward366** [comment]
I see what you mean for this control in a table.  That's fine.  We can leave it out for now.

File: canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx [hunk 1]
```diff
@@ -45,7 +46,7 @@ class TextfieldControl extends React.Component {
 		if (this.charLimit !== -1 && value) {
 			value = value.substring(0, this.charLimit);
 		}
-		if (this.isList) {
+		if (this.isList && this.props.control.controlType !== ControlType.LIST) {
```
[the remark this answers is not in the document]
Comment (caritaou) [reply canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx:49]: This textfield inside a List control is of type `list`. The first example in the list_paramDef, the propertyId is `{ name: "list_string", row: 3}`.

If we remove this check, if the input contains `, `, the string will get parsed into an array:
image

File: canvas_modules/common-canvas/src/common-properties/properties-controller.js [hunk 2]
```diff
@@ -1236,6 +1236,14 @@ export default class PropertiesController {
 		return control;
 	}
 
+	getControlPropType(propertyId) {
```
[the remark this answers is not in the document]
Comment (caritaou) [reply canvas_modules/common-canvas/src/common-properties/properties-controller.js:1239]: added tests to properties-controller-test.js

File: canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/textfield/textfield.jsx:49]: Something is wrong here.  This control should not be a list.  The control should think its type is a string.  We shouldn't be sending down the parent's control object.

File: canvas_modules/common-canvas/src/common-properties/controls/list/list.jsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+	}
+
+	makeRows(controlValue, tableState) {
+		const rows = [];
+		if (controlValue) {
+			for (var rowIndex = 0; rowIndex < controlValue.length; rowIndex++) {
+				const columns = [];
+				const row = { row: rowIndex };
+				const propertyId = Object.assign({}, this.props.propertyId, row);
+				const control = Object.assign({}, this.props.control);
```
Comment (caritaou) [inline canvas_modules/common-canvas/src/common-properties/controls/list/list.jsx:75]: @matthoward366  continuing the thread from https://github.com/elyra-ai/canvas/pull/171/files#r458109412
The list control is creating its own rows of TextfieldControl/NumberfieldControl passing in its own information. Unlike other complexType controls, we don't have parameter information to create the control from.

Should we use default values for the control here? In the case of textfields, something like 
```
{
  controlType: "textfield",
  editStyle: "inline",
  valueDef: {
    isList: false
    isMap: false
    propType: "string"
  }
}
```

Comment (matthoward366) [reply canvas_modules/common-canvas/src/common-properties/controls/list/list.jsx:75]: Can we just pass down an empty control?  Something like control={}.  It looks like the textfield has checks that default correctly.

**Comment by caritaou** [comment]
Theres a cypress test failing, I need to rebase with master after https://github.com/elyra-ai/canvas/issues/176 is fixed

**Pull request merged**